api/handlers.go

```diff
@@ -439,15 +446,24 @@ func (api *API) userLimitsGET(_ *database.User, w http.ResponseWriter, req *http
 			return
 		}
 		api.staticUserTierCache.Set(u)
-	}
-	tier, ok = api.staticUserTierCache.Get(sub)
-	if !ok {
-		build.Critical("Failed to fetch user from UserTierCache right after setting it.")
+		// Populate the tier and qe values, while simultaneously making sure
+		// that we can read the record from the cache.
+		tier, qe, ok = api.staticUserTierCache.Get(sub)
+		if !ok {
+			build.Critical("Failed to fetch user from UserTierCache right after setting it.")
+		}
 	}
 	resp := UserLimitsGET{
 		TierID:     tier,
 		TierLimits: database.UserLimits[tier],
 	}
+	// If the quota is exceeded we should keep the user's tier but report
+	// anonymous-level speeds.
+	if qe {
+		resp.TierLimits = database.UserLimits[database.TierAnonymous]
+		// Keep the original tier name.
+		resp.TierLimits.TierName = database.UserLimits[tier].TierName
```

Comment:
Right, this is not great. `TierName` is always reported - it's part of the `database.TierLimits` struct that makes up `api.UserLimitsGET`. The funky thing happening here is that we need to replace the entire struct we return but we also need to keep the tier and tier name as the original user's tier and tier name. Because they still have that tier (i.e. they can access pro portals) but they are temporarily limited to lower speeds.

I refactored it a bit. `UserLimitsGET` now explicitly lists all fields that make it up and there's a helper method for creating it. I can't help the clunky 
```go
		if u.QuotaExceeded {
			// Report the speeds for tier anonymous.
			resp = userLimitsGetFromTier(database.TierAnonymous)
			// But keep reporting the user's actual tier and it's name.
			resp.TierID = u.Tier
			resp.TierName = database.UserLimits[u.Tier].TierName
		}
```
but I don't think there's a way to get around that.
If you have an idea - I'm open to suggestions.

Reply:
that's fine for me I was just under the impression the `TierName` was not always listed in the report. As long as all of the properties in the response are always there I think it's better.